applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h

```diff
@@ -454,6 +599,33 @@ class GeometryUtilities
         }
     }
 
+    // --------------------------------------------------------------------------
+    SizeType GetNumberOfElementalNodesInModelPart(const ModelPart& rModelPart) const
+    {
+        KRATOS_TRY
+
+        const int local_geometry_type = block_for_each<MaxReduction<int>>(mrModelPart.Elements(), [&](const ModelPart::ElementType& rElement) {
+            return static_cast<int>(rElement.GetGeometry().GetGeometryType());
+        });
+        const int global_geometry_type = mrModelPart.GetCommunicator().GetDataCommunicator().MaxAll(local_geometry_type);
+
+        block_for_each(mrModelPart.Elements(), [&](const ModelPart::ElementType& rElement) {
+            KRATOS_ERROR_IF(static_cast<int>(rElement.GetGeometry().GetGeometryType()) != global_geometry_type)
```

Comment:
I dont get the purpose of this function:

why is it required that all elements are of the same type? simply summing up the nodes of each element nodes should work independently of element types. 

Additionally: summing up the number of nodes will count many nodes multiple times.

Reply:
> why is it required that all elements are of the same type? simply summing up the nodes of each element nodes should work independently of element types.

It is required to have same `Element` type in the sensitivity calculation because, I don't want to check element type in the OMP parallelized for loop using if blocks since, it will creat lots of branches within for loop which is inefficient. Following if block needs to be executed for each element if this check is not performed.

https://github.com/KratosMultiphysics/Kratos/blob/fb02002c19360ebfa44c0ee98832feae2f7e35c8/applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h#L399-L437

I wanted to avoid that.

> Additionally: summing up the number of nodes will count many nodes multiple times.

It is not summing up, it is taking the `Max` out of all.

Reply:
Ok i got why you need to do this here, thanks for explaining.

I am of the opinion that we should pay the price of having those "if" branches in the omp loop, since that will allow to use any type of mesh and combination of elements. 

If the the volume derivative is not implemented in the element/geometry interface, that is the price we have to pay. 
For such a tool like a dedicated response function, i think it is more important to work on generic meshes rather to be maximum efficient (especially since this is cheap anyway.

Reply:
I will change accordingly

Reply:
done :)